age-piv-p256.md

```diff
@@ -0,0 +1,100 @@
+# age piv-p256 recipient stanza
+
+[c2sp.org/age-piv-p256](https://c2sp.org/age-piv-p256)
+
+## Introduction
+
+This document specifies the age piv-p256 recipient stanza. This stanza
+is used for encryption using NIST P-256 keys stored on hardware tokens
+or PIV devices.
+
+Apart from using the NIST P-256 curve, the piv-p256 recipient stanza
+differs from the native [X25519 recipient
+stanza](https://age-encryption.org/v1#x25519-recipient-stanza) in that
+it includes a tag of the recipient, to allow the recipient to select the
+correct secret without the hardware token requesting user interaction.
+
+## Conventions used in this document
+
+ABNF syntax follows [RFC
+5234](https://www.rfc-editor.org/rfc/rfc5234.html) and [RFC
+7405](https://www.rfc-editor.org/rfc/rfc7405.html) and references the
+core rules in RFC 5234, Appendix B.1.
+
+The base64 encoding used throughout is the standard Base 64 encoding
+specified in [RFC 4648](https://www.rfc-editor.org/rfc/rfc4648.html),
+Section 4, without `=` padding characters (sometimes referred to as
+"raw" or "unpadded" base64). Encoders MUST generate canonical base64
+according to RFC 4648, Section 3.5, and decoders MUST reject
+non-canonical encodings and encodings ending with `=` padding
+characters.
+
+`||` denotes concatenation. `0x` followed by two hexadecimal characters
+denotes a byte value in the 0-255 range.
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and
+"OPTIONAL" in this document are to be interpreted as described in [BCP
+14](https://www.rfc-editor.org/info/bcp14) [RFC
+2119](https://www.rfc-editor.org/rfc/rfc2119.html) [RFC
+8174](https://www.rfc-editor.org/rfc/rfc8174.html) when, and only when,
+they appear in all capitals, as shown here.
+
+## Recipient stanza
+
+A piv-p256 recipient stanza has three arguments.
+
+    -> piv-p256 8u8pZg Aqa5k2Tm/BJGTzCno1W29dWaWBHkFchOztcIZ7aQo0Ca
+    CXE0N0KMI0BUE5WaYTLOxYv7aHB/5IQpAZKsk2C/yjU
+
+1.  The first argument is the fixed string `piv-p256`.
+
+2.  The second argument is the base-64 encdoded tag of the recipient,
+    consisting of the first 4 bytes of the SHA-256 hash of the
+    recipient:
+
+         tag = SHA-256(recipient)[0..4]
```

Comment:
I was wondering if the tag should be salted, to hide relations between stanzas (in different files) for the same recipient

Reply:
Hmm what do you have in mind?

Maybe `x = CSPRNG(4); tag = x || HMAC(recipient, x)[0..4]` so that only someone who knows the recipient can link files?

That is indeed a nice step forward in privacy! Might be worth making a backwards-incompatible change. @str4d, what do you think?

Reply:
I think it's pretty common to just use the x-coordinate, similar to X25519

Reply:
> I think it's pretty common to just use the x-coordinate, similar to X25519

For the shared ECDH output yes, but the input is usually either a compressed or an uncompressed point, no? x-coordinate-only ECDH like in X25519 is very uncommon on P-256. (I recently surveyed a bunch of ECDH and ECIES implementations while designing the crypto/ecdh.)

Reply:
Which spec uses a SEC1-encoded point as the input?

TLS 1.3 (by way of IEEE1363) uses the x-coordinate, for example:

https://datatracker.ietf.org/doc/html/rfc8446#section-7.4.2

> [7.4.2](https://datatracker.ietf.org/doc/html/rfc8446#section-7.4.2).  Elliptic Curve Diffie-Hellman
>
>   For secp256r1, secp384r1, and secp521r1, ECDH calculations (including
>   parameter and key generation as well as the shared secret
>   calculation) are performed according to [[IEEE1363](https://datatracker.ietf.org/doc/html/rfc8446#ref-IEEE1363)] using the
>   ECKAS-DH1 scheme with the identity map as the key derivation function
>   (KDF), so that the shared secret is the x-coordinate of the ECDH
>   shared secret elliptic curve point represented as an octet string.

Reply:
My bad, I was originally trying to reply to [this comment](https://github.com/C2SP/C2SP/pull/31/files/4c6c0ad4a6993a83d635a2e74602d4128c3e9ce3#r1305437492)

Reply:
> Maybe `x = CSPRNG(4); tag = x || HMAC(recipient, x)[0..4]` so that only someone who knows the recipient can link files?

That's indeed what I had in mind, except I was thinking of putting `x` in a separate stanza argument instead of concatenating it (to avoid a custom split)

Reply:
Is there a reason to prefer this construction instead of `HMAC(recipient, ephemeral share)[0..4]`?

Reply:
IANAC, but sounds nice and simple to me.